metalearn/metafeatures/metafeatures.py

```diff
@@ -86,57 +84,81 @@ def compute(
         one for the value and one for the compute time of that metafeature
         value
         """
-        self.computed_metafeatures = DataFrame()
-        if timeout is None:
-            self._compute(
+        if timeout is not None:
+            timeout = timeout - self.TIMEOUT_BUFFER
+
+        self._threadsafe_timeout_function(
+            self._compute,
+            (
                 X, Y, column_types, metafeature_ids, sample_rows,
                 sample_columns, seed
-            )
-        else:
-            with redirect_stderr(io.StringIO()):
-                threadsafe_timeout_function(
-                    self._compute,
-                    (
-                        X, Y, column_types, metafeature_ids, sample_rows,
-                        sample_columns, seed
-                    ),
-                    timeout-self.TIMEOUT_BUFFER
-                )
+            ),
+            timeout,
+            metafeature_ids
+        )
+
+        if not self.queue.empty():
+            self.computed_metafeatures = self.queue.get()
+            for x in range(self.queue.qsize()):
+                mf, value = self.queue.get()
+                self.computed_metafeatures.at[0, mf] = value
+
         return self.computed_metafeatures
 
+    def _threadsafe_timeout_function(self, f, args, timeout, metafeature_ids):
+        p = multiprocessing.Process(target=f, args=args)
+        p.start()
+        try:
+            p.join(timeout)
+            if p.is_alive():
+                p.terminate()
+                p.join()
+        except multiprocessing.TimeoutError:
+            print("timeout")        
```

Comment:
we should not print anything on timeout, just return the TIMEOUT flag for any metafeatures not computed